Changelog.md

```diff
@@ -14,6 +14,7 @@ Compiler Features:
 Bugfixes:
  * Commandline Interface: Disallow the following options outside of the compiler mode: ``--via-ir``,``--metadata-literal``, ``--metadata-hash``, ``--model-checker-show-unproved``, ``--model-checker-div-mod-no-slacks``, ``--model-checker-engine``, ``--model-checker-invariants``, ``--model-checker-solvers``, ``--model-checker-timeout``, ``--model-checker-contracts``, ``--model-checker-targets``.
  * Type Checker: Fix null dereference in `abi.encodeCall` type checking of free function.
+ * Type Checker: Fix compiler crashing due to out-of-range error when left-hand side tuples are nested in outer unary tuple during tuple assignment.
```

Comment:
```suggestion
 * Type Checker: Fix compiler crash when left-hand side tuples are nested in outer unary tuple during tuple assignment.
```
I think that we could remove the reason of the compiler crash since it is cpp related and not useful for end-user. 
Also, it seems that the problem does not happen, when the nested tuple is the first element:
```
     function f() public pure {
        ((((, ())), )) = (1, 3);
    }
    // out of range crash
    function g() public pure {
        ((, ((, )))) = (1, 3);
    }
```
However, if we add an empty tuple to `function f()` above, we get the crash again:
```
     // Not ok anymore
     function f() public pure {
        ((((, ())), ())) = (1, 3);
    }
```
So, I am not sure if this is strictly related to tuples nested in outer unary tuple... I guess we could investigate it a little bit more to understand the exact situation which triggers the crash.

Reply:
I'll amend the changelog.

Here's my understanding of the cases you pointed out:
When the nested tuple is in the first element, the crash is normally due to the call `_lhs.components().at(index)`. Here, `index = 0`. This doesn't cause a crash because `_lhs.components()` will always have at least 1 element. This is a lucky coincidence. 

But the element `index = 0` refers to is not `_lhs.components().at(0)` but rather `lhsType.components().at(0)`. When we retrieve the type, the extra meaningless nested unary tuples are already removed so these two elements are not the same. This crash only reveals itself when there's a mismatch between the sizes of `_lhs` and `lhsType`.
```solidity
function f() public pure {
    ((((, ())), )) = (1, 3);
}
```

Reply:
Seems like my answer isn't quite correct.

This is not fine:
```solidity
function f() pure {
    ((, ())) = (1,3);
}
```
But this is fine:
```solidity
function f() pure {
    ((,, ())) = (1,3);
}
```
I'll look into it more

Reply:
I think the gist is that the bug could be triggered by any recursive unary tuple, not just an "outer" one.

Reply:
I think I understood it correctly. The reason why the below works is because it failed the size check at 

https://github.com/ethereum/solidity/blob/d5a78b18b3fd9e54b2839e9685127c6cdbddf614/libsolidity/analysis/TypeChecker.cpp#L116

```solidity
// This fails because lhsType.size() == _rhs.size()
function f() pure {
    ((, ())) = (1,3);
}

// This is ok because the size check caught size mismatch before the crash bug
// lhsType.size() == 3
// _rhs.size() == 2
function f() pure {
    ((,, ())) = (1,3);
}

// if we guarantee the sizes are the same, then crash happens
function f() pure {
    ((,, ())) = (1,3,4);
}
```

Reply:
> I think the gist is that the bug could be triggered by any recursive unary tuple, not just an "outer" one (it does happen during a recursion into the tuples after all).

Right. I agree.

Reply:
I'm trying to understand why this happens:
```solidity
// works
function f() pure {
    ((((, ())) , )) = ((1, 2), 3);
}

// if the outer unary tuple is taken out, it crashes again
// this is the opposite of the simple cases where adding outer unary tuple causes the crash
function g() pure {
    (((, ())) , ) = ((1, 2), 3);
}
```

Reply:
So far the fix has addressed all the cases. But yea I changed the phrasing of the changelog to be more accurate.